Index: gnome-menus-3.7.90/po/POTFILES.in
===================================================================
--- gnome-menus-3.7.90.orig/po/POTFILES.in	2013-02-19 18:39:14.000000000 -0500
+++ gnome-menus-3.7.90/po/POTFILES.in	2013-03-03 00:59:16.629486723 -0500
@@ -15,3 +15,19 @@
 desktop-directories/X-GNOME-Sundry.directory.in
 desktop-directories/X-GNOME-Utilities.directory.in
 desktop-directories/X-GNOME-WebApplications.directory.in
+debian/desktop-files/ActionGames.directory.in
+debian/desktop-files/AdventureGames.directory.in
+debian/desktop-files/ArcadeGames.directory.in
+debian/desktop-files/BlocksGames.directory.in
+debian/desktop-files/BoardGames.directory.in
+debian/desktop-files/CardGames.directory.in
+debian/desktop-files/Debian.directory.in
+debian/desktop-files/GnomeScience.directory.in
+debian/desktop-files/KidsGames.directory.in
+debian/desktop-files/LogicGames.directory.in
+debian/desktop-files/RolePlayingGames.directory.in
+debian/desktop-files/Settings-System.directory.in
+debian/desktop-files/Settings.directory.in
+debian/desktop-files/SimulationGames.directory.in
+debian/desktop-files/SportsGames.directory.in
+debian/desktop-files/StrategyGames.directory.in
